File: ios/XiangshangYouth/Features/Parent/CaptureCalibration.swift
import SwiftUI
import CoreMotion
import Combine

struct CaptureDeviceAlignment: Equatable {
    let rollDegrees: Double
    let pitchDegrees: Double
    let isAvailable: Bool

    static let pending = CaptureDeviceAlignment(rollDegrees: 0, pitchDegrees: 0, isAvailable: false)

    var isLevel: Bool {
        isAvailable && abs(rollDegrees) <= CaptureCalibrationRules.maximumRollDegrees
            && abs(pitchDegrees) <= CaptureCalibrationRules.maximumPitchDegrees
    }
}

struct CaptureBodyAlignment: Equatable {
    let bodyDetected: Bool
    let distanceState: BodyCaptureQualityGate.BodyScaleState
    let centered: Bool
    let headReady: Bool
    let shouldersReady: Bool
    let hipsReady: Bool
    let kneesReady: Bool
    let feetReady: Bool

    static let waiting = CaptureBodyAlignment(
        bodyDetected: false,
        distanceState: .invalid,
        centered: false,
        headReady: false,
        shouldersReady: false,
        hipsReady: false,
        kneesReady: false,
        feetReady: false
    )

    var distanceReady: Bool { distanceState == .ready }
    var isReady: Bool {
        bodyDetected && distanceReady && centered && headReady && shouldersReady && hipsReady && kneesReady && feetReady
    }
}

enum CaptureCalibrationRules {
    /// Commercial fixed-pose capture uses a deliberately tighter gate than
    /// ordinary video guidance. These bounds gate collection only; they do
    /// not make a medical or accuracy claim.
    static let maximumRollDegrees = 2.0
    static let maximumPitchDegrees = 5.0

    static func deviceAlignment(gravityX: Double, gravityY: Double, gravityZ: Double) -> CaptureDeviceAlignment {
        guard [gravityX, gravityY, gravityZ].allSatisfy(\.isFinite) else { return .pending }
        let magnitude = sqrt(gravityX * gravityX + gravityY * gravityY + gravityZ * gravityZ)
        guard magnitude >= 0.5 else { return .pending }
        let x = gravityX / magnitude
        let y = gravityY / magnitude
        let z = gravityZ / magnitude
        let roll = atan2(x, -y) * 180 / .pi
        let pitch = atan2(z, max(hypot(x, y), 0.001)) * 180 / .pi
        return CaptureDeviceAlignment(rollDegrees: roll, pitchDegrees: pitch, isAvailable: true)
    }
}

@MainActor
final class CaptureDeviceAlignmentMonitor: ObservableObject {
    @Published private(set) var alignment: CaptureDeviceAlignment = .pending
    private let motion = CMMotionManager()

    func start() {
        guard !motion.isDeviceMotionActive else { return }
        motion.deviceMotionUpdateInterval = 1.0 / 20.0
        guard motion.isDeviceMotionAvailable else {
            alignment = .pending
            return
        }
        motion.startDeviceMotionUpdates(to: .main) { [weak self] data, _ in
            guard let self, let gravity = data?.gravity else { return }
            let next = CaptureCalibrationRules.deviceAlignment(
                gravityX: gravity.x,
                gravityY: gravity.y,
                gravityZ: gravity.z
            )
            guard next.isAvailable else { return }
            if self.alignment.isAvailable {
                self.alignment = CaptureDeviceAlignment(
                    rollDegrees: self.alignment.rollDegrees * 0.72 + next.rollDegrees * 0.28,
                    pitchDegrees: self.alignment.pitchDegrees * 0.72 + next.pitchDegrees * 0.28,
                    isAvailable: true
                )
            } else {
                self.alignment = next
            }
        }
    }

    func stop() {
        motion.stopDeviceMotionUpdates()
        alignment = .pending
    }

    deinit { motion.stopDeviceMotionUpdates() }
}

struct CaptureHumanCalibrationGuide: View {
    let task: BodyAssessmentRecord.CaptureTask
    let alignment: CaptureBodyAlignment
    let device: CaptureDeviceAlignment
    let recording: Bool

    var body: some View {
        GeometryReader { proxy in
            let guideHeight = min(proxy.size.height * 0.46, task == .footArch ? 300 : (task == .seatedPosture ? 300 : 430))
            let guideWidth = min(proxy.size.width * (task == .footArch ? 0.78 : 0.66), task == .seatedPosture ? 280 : 300)
            VStack(spacing: 10) {
                DeviceLevelBubble(alignment: device)
                Canvas { context, size in
                    let baseColor = alignment.isReady ? Color.green : Color.white.opacity(0.82)
                    let muted = Color.white.opacity(0.34)
                    func point(_ x: CGFloat, _ y: CGFloat) -> CGPoint { CGPoint(x: size.width * x, y: size.height * y) }
                    func line(_ a: CGPoint, _ b: CGPoint, color: Color? = nil, width: CGFloat = 3) {
                        var path = Path(); path.move(to: a); path.addLine(to: b)
                        context.stroke(path, with: .color(color ?? baseColor), style: StrokeStyle(lineWidth: width, lineCap: .round))
                    }
                    if task == .footArch {
                        let leftKnee = point(0.34, 0.14), rightKnee = point(0.66, 0.14)
                        let leftAnkle = point(0.36, 0.72), rightAnkle = point(0.64, 0.72)
                        let leftHeel = point(0.31, 0.84), rightHeel = point(0.59, 0.84)
                        let leftToe = point(0.47, 0.90), rightToe = point(0.75, 0.90)
                        line(leftKnee, leftAnkle, color: alignment.kneesReady ? .green : baseColor, width: 5)
                        line(rightKnee, rightAnkle, color: alignment.kneesReady ? .green : baseColor, width: 5)
                        line(leftAnkle, leftHeel, color: alignment.feetReady ? .green : baseColor, width: 5)
                        line(leftHeel, leftToe, color: alignment.feetReady ? .green : baseColor, width: 5)
                        line(rightAnkle, rightHeel, color: alignment.feetReady ? .green : baseColor, width: 5)
                        line(rightHeel, rightToe, color: alignment.feetReady ? .green : baseColor, width: 5)
                        for pointValue in [leftKnee, rightKnee, leftAnkle, rightAnkle, leftHeel, rightHeel, leftToe, rightToe] {
                            context.fill(Path(ellipseIn: CGRect(x: pointValue.x - 6, y: pointValue.y - 6, width: 12, height: 12)), with: .color(alignment.isReady ? .green : .yellow))
                        }
                    } else {
                    let showsAdamsPose = task == .forwardBend && recording
                    let head = point(0.5, task == .seatedPosture ? 0.20 : (showsAdamsPose ? 0.43 : 0.12))
                    let shoulderY: CGFloat = task == .seatedPosture ? 0.38 : (showsAdamsPose ? 0.49 : 0.30)
                    let hipY: CGFloat = task == .seatedPosture ? 0.68 : 0.55
                    let footY: CGFloat = 0.91
                    let leftShoulder = point(0.34, shoulderY), rightShoulder = point(0.66, shoulderY)
                    let leftHip = point(0.42, hipY), rightHip = point(0.58, hipY)
                    let isAdams = task == .forwardBend
                    let leftKnee = point(isAdams ? 0.47 : 0.40, task == .seatedPosture ? 0.82 : 0.73)
                    let rightKnee = point(isAdams ? 0.53 : 0.60, task == .seatedPosture ? 0.82 : 0.73)
                    let leftFoot = point(isAdams ? 0.47 : 0.39, footY), rightFoot = point(isAdams ? 0.53 : 0.61, footY)
                    let headRadius = min(size.width, size.height) * 0.07
                    context.stroke(Path(ellipseIn: CGRect(x: head.x - headRadius, y: head.y - headRadius, width: headRadius * 2, height: headRadius * 2)), with: .color(alignment.headReady ? .green : baseColor), lineWidth: 3)
                    line(point(0.5, head.y / size.height + (showsAdamsPose ? 0.025 : 0.07)), point(0.5, hipY), color: alignment.centered ? .green : baseColor)
                    line(leftShoulder, rightShoulder, color: alignment.shouldersReady ? .green : baseColor)
                    line(leftShoulder, point(isAdams ? 0.49 : 0.24, isAdams ? (showsAdamsPose ? 0.70 : 0.50) : (task == .seatedPosture ? 0.58 : 0.50)), color: alignment.shouldersReady ? .green : baseColor)
                    line(rightShoulder, point(isAdams ? 0.51 : 0.76, isAdams ? (showsAdamsPose ? 0.70 : 0.50) : (task == .seatedPosture ? 0.58 : 0.50)), color: alignment.shouldersReady ? .green : baseColor)
                    line(leftHip, rightHip, color: alignment.hipsReady ? .green : baseColor)
                    if task != .seatedPosture {
                        line(leftHip, leftKnee, color: alignment.kneesReady ? .green : baseColor)
                        line(rightHip, rightKnee, color: alignment.kneesReady ? .green : baseColor)
                        line(leftKnee, leftFoot, color: alignment.feetReady ? .green : baseColor)
                        line(rightKnee, rightFoot, color: alignment.feetReady ? .green : baseColor)
                    } else {
                        line(leftHip, point(0.35, 0.88), color: muted)
                        line(rightHip, point(0.65, 0.88), color: muted)
                    }
                    for (pointValue, ready) in [(head, alignment.headReady), (leftShoulder, alignment.shouldersReady), (rightShoulder, alignment.shouldersReady), (leftHip, alignment.hipsReady), (rightHip, alignment.hipsReady), (leftKnee, alignment.kneesReady), (rightKnee, alignment.kneesReady), (leftFoot, alignment.feetReady), (rightFoot, alignment.feetReady)] {
                        context.fill(Path(ellipseIn: CGRect(x: pointValue.x - 5, y: pointValue.y - 5, width: 10, height: 10)), with: .color(ready ? .green : .yellow))
                    }
                    }
                }
                .opacity(recording ? 0.42 : 1)
                .overlay {
                    RoundedRectangle(cornerRadius: 34)
                        .stroke(alignment.distanceReady ? Color.green.opacity(0.9) : Color.white.opacity(0.72), style: StrokeStyle(lineWidth: 2, dash: [9, 7]))
                }
                .frame(width: guideWidth, height: guideHeight)
                Text(task == .footArch
                     ? (recording ? "正在记录足部近景：保持双脚平行站稳" : "仅将双膝以下、双踝、足跟和双脚放入近景框")
                     : (recording
                        ? (task == .forwardBend ? "对准前屈轮廓：双脚并拢、双膝伸直、头部放松" : "正在记录：保持手机固定，按语音完成动作")
                        : "将头、肩、髋、膝和双脚对准标定点"))
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(alignment.isReady ? .green : .white)
                    .padding(.horizontal, 12).padding(.vertical, 7)
                    .background(.black.opacity(0.50), in: Capsule())
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .offset(y: -44)
        }
        .ignoresSafeArea()
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("采集标定。手机\(device.isLevel ? "已校正" : "未校正")，\(task == .footArch ? "足部近景" : "人体")\(alignment.isReady ? "已对齐" : "未完全对齐")")
    }
}

private struct DeviceLevelBubble: View {
    let alignment: CaptureDeviceAlignment

    var body: some View {
        HStack(spacing: 8) {
            ZStack {
                Circle().stroke(.white.opacity(0.55), lineWidth: 1)
                Circle().fill(alignment.isLevel ? .green : .yellow)
                    .frame(width: 9, height: 9)
                    .offset(
                        x: CGFloat(min(max(alignment.rollDegrees, -10), 10)) * 1.2,
                        y: CGFloat(min(max(alignment.pitchDegrees, -10), 10)) * 1.2
                    )
                Rectangle().fill(.white.opacity(0.45)).frame(width: 1, height: 28)
                Rectangle().fill(.white.opacity(0.45)).frame(width: 28, height: 1)
            }
            .frame(width: 34, height: 34)
            VStack(alignment: .leading, spacing: 1) {
                Text(alignment.isLevel ? "手机角度已校正" : "请调平手机")
                    .font(.system(size: 13, weight: .bold))
                Text(alignment.isAvailable ? "左右 \(Int(alignment.rollDegrees.rounded()))° · 前后 \(Int(alignment.pitchDegrees.rounded()))°" : "正在读取陀螺仪")
                    .font(.system(size: 12, weight: .medium)).opacity(0.75)
            }
        }
        .foregroundStyle(.white)
        .padding(.horizontal, 12).padding(.vertical, 7)
        .background(.black.opacity(0.50), in: Capsule())
    }
}

struct CaptureCalibrationStatusRow: View {
    let cameraReady: Bool
    let deviceReady: Bool
    let bodyReady: Bool

    var body: some View {
        HStack(spacing: 6) {
            item("相机", ready: cameraReady, icon: "camera.fill")
            item("手机角度", ready: deviceReady, icon: "gyroscope")
            item("人体对齐", ready: bodyReady, icon: "figure.stand")
        }
    }

    private func item(_ title: String, ready: Bool, icon: String) -> some View {
        Label(title, systemImage: ready ? "checkmark.circle.fill" : icon)
            .font(.system(size: 12, weight: .semibold))
            .foregroundStyle(ready ? .green : .white.opacity(0.72))
            .frame(maxWidth: .infinity, minHeight: 30)
            .background(.white.opacity(0.07), in: Capsule())
    }
}

File: android/app/src/main/java/com/xiangshang/youth/feature/parent/CaptureCalibration.kt
package com.xiangshang.youth.feature.parent

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CameraAlt
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.ScreenRotation
import androidx.compose.material.icons.filled.AccessibilityNew
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.xiangshang.youth.core.model.BodyCaptureQualityGate
import com.xiangshang.youth.core.model.BodyCaptureTask
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.hypot
import kotlin.math.sqrt

data class CaptureDeviceAlignment(
    val rollDegrees: Double = 0.0,
    val pitchDegrees: Double = 0.0,
    val available: Boolean = false
) {
    val isLevel: Boolean get() = available && abs(rollDegrees) <= CaptureCalibrationRules.maximumRollDegrees && abs(pitchDegrees) <= CaptureCalibrationRules.maximumPitchDegrees
}

data class CaptureBodyAlignment(
    val bodyDetected: Boolean = false,
    val distanceState: BodyCaptureQualityGate.BodyScaleState = BodyCaptureQualityGate.BodyScaleState.Invalid,
    val centered: Boolean = false,
    val headReady: Boolean = false,
    val shouldersReady: Boolean = false,
    val hipsReady: Boolean = false,
    val kneesReady: Boolean = false,
    val feetReady: Boolean = false
) {
    val distanceReady: Boolean get() = distanceState == BodyCaptureQualityGate.BodyScaleState.Ready
    val isReady: Boolean get() = bodyDetected && distanceReady && centered && headReady && shouldersReady && hipsReady && kneesReady && feetReady
}

object CaptureCalibrationRules {
    /** Initial commercial fixed-pose capture bounds. They gate collection;
     * they do not make a medical or accuracy claim. */
    const val maximumRollDegrees = 2.0
    const val maximumPitchDegrees = 5.0

    fun deviceAlignment(gravityX: Double, gravityY: Double, gravityZ: Double): CaptureDeviceAlignment {
        if (!listOf(gravityX, gravityY, gravityZ).all(Double::isFinite)) return CaptureDeviceAlignment()
        val magnitude = sqrt(gravityX * gravityX + gravityY * gravityY + gravityZ * gravityZ)
        if (magnitude < .5) return CaptureDeviceAlignment()
        val x = gravityX / magnitude
        val y = gravityY / magnitude
        val z = gravityZ / magnitude
        val roll = Math.toDegrees(atan2(x, y))
        val pitch = Math.toDegrees(atan2(z, hypot(x, y).coerceAtLeast(.001)))
        return CaptureDeviceAlignment(roll, pitch, true)
    }
}

@Composable
fun rememberCaptureDeviceAlignment(): State<CaptureDeviceAlignment> {
    val context = LocalContext.current
    val state = remember { mutableStateOf(CaptureDeviceAlignment()) }
    DisposableEffect(context) {
        val manager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        val sensor = manager.getDefaultSensor(Sensor.TYPE_GRAVITY) ?: manager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        val listener = object : SensorEventListener {
            override fun onSensorChanged(event: SensorEvent) {
                val next = CaptureCalibrationRules.deviceAlignment(event.values[0].toDouble(), event.values[1].toDouble(), event.values[2].toDouble())
                if (!next.available) return
                val old = state.value
                state.value = if (old.available) {
                    CaptureDeviceAlignment(old.rollDegrees * .72 + next.rollDegrees * .28, old.pitchDegrees * .72 + next.pitchDegrees * .28, true)
                } else next
            }
            override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) = Unit
        }
        if (sensor != null) manager.registerListener(listener, sensor, SensorManager.SENSOR_DELAY_GAME)
        onDispose { manager.unregisterListener(listener) }
    }
    return state
}

@Composable
fun CaptureHumanCalibrationGuide(task: BodyCaptureTask, body: CaptureBodyAlignment, device: CaptureDeviceAlignment, recording: Boolean) {
    BoxWithConstraints(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        val guideWidth = minOf(maxWidth * if (task == BodyCaptureTask.FootArch) .78f else .66f, if (task == BodyCaptureTask.Seated) 280.dp else 300.dp)
        val guideHeight = minOf(maxHeight * .46f, if (task == BodyCaptureTask.FootArch || task == BodyCaptureTask.Seated) 300.dp else 430.dp)
        Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(10.dp), modifier = Modifier.offset(y = (-42).dp)) {
            DeviceLevelBubble(device)
            Canvas(
                Modifier
                    .width(guideWidth)
                    .height(guideHeight)
                    .alpha(if (recording) .42f else 1f)
                    .semantics { contentDescription = if (task == BodyCaptureTask.FootArch) "足部近景框，${if (body.isReady) "已经对齐" else "等待膝踝足跟对齐"}" else "人体对齐框，${if (body.isReady) "已经对齐" else "等待头肩髋脚对齐"}" }
            ) {
                val base = if (body.isReady) Color(0xFF28C76F) else Color.White.copy(alpha = .84f)
                if (task == BodyCaptureTask.FootArch) {
                    val lk = Offset(size.width * .34f, size.height * .14f); val rk = Offset(size.width * .66f, size.height * .14f)
                    val la = Offset(size.width * .36f, size.height * .72f); val ra = Offset(size.width * .64f, size.height * .72f)
                    val lh = Offset(size.width * .31f, size.height * .84f); val rh = Offset(size.width * .59f, size.height * .84f)
                    val lt = Offset(size.width * .47f, size.height * .90f); val rt = Offset(size.width * .75f, size.height * .90f)
                    fun footLine(a: Offset, b: Offset, ready: Boolean) = drawLine(if (ready) Color(0xFF28C76F) else base, a, b, 5.dp.toPx(), StrokeCap.Round)
                    drawRoundRect(if (body.distanceReady) Color(0xFF28C76F) else Color.White.copy(alpha = .72f), style = Stroke(2.dp.toPx()), cornerRadius = androidx.compose.ui.geometry.CornerRadius(34.dp.toPx()))
                    footLine(lk, la, body.kneesReady); footLine(rk, ra, body.kneesReady)
                    footLine(la, lh, body.feetReady); footLine(lh, lt, body.feetReady)
                    footLine(ra, rh, body.feetReady); footLine(rh, rt, body.feetReady)
                    listOf(lk, rk, la, ra, lh, rh, lt, rt).forEach { point -> drawCircle(if (body.isReady) Color(0xFF28C76F) else Color(0xFFFFD54F), 6.dp.toPx(), point) }
                } else {
                val showsAdamsPose = task == BodyCaptureTask.ForwardBend && recording
                val shoulderY = when {
                    task == BodyCaptureTask.Seated -> size.height * .38f
                    showsAdamsPose -> size.height * .49f
                    else -> size.height * .30f
                }
                val hipY = if (task == BodyCaptureTask.Seated) size.height * .68f else size.height * .55f
                val head = Offset(size.width * .5f, size.height * when {
                    task == BodyCaptureTask.Seated -> .20f
                    showsAdamsPose -> .43f
                    else -> .12f
                })
                val ls = Offset(size.width * .34f, shoulderY); val rs = Offset(size.width * .66f, shoulderY)
                val lh = Offset(size.width * .42f, hipY); val rh = Offset(size.width * .58f, hipY)
                val isAdams = task == BodyCaptureTask.ForwardBend
                val lk = Offset(size.width * if (isAdams) .47f else .40f, size.height * if (task == BodyCaptureTask.Seated) .82f else .73f)
                val rk = Offset(size.width * if (isAdams) .53f else .60f, size.height * if (task == BodyCaptureTask.Seated) .82f else .73f)
                val lf = Offset(size.width * if (isAdams) .47f else .39f, size.height * .91f); val rf = Offset(size.width * if (isAdams) .53f else .61f, size.height * .91f)
                fun line(a: Offset, b: Offset, ready: Boolean) = drawLine(if (ready) Color(0xFF28C76F) else base, a, b, 3.dp.toPx(), StrokeCap.Round)
                drawRoundRect(if (body.distanceReady) Color(0xFF28C76F) else Color.White.copy(alpha = .72f), style = Stroke(2.dp.toPx()), cornerRadius = androidx.compose.ui.geometry.CornerRadius(34.dp.toPx()))
                drawCircle(if (body.headReady) Color(0xFF28C76F) else base, size.minDimension * .07f, head, style = Stroke(3.dp.toPx()))
                line(Offset(size.width * .5f, head.y + size.height * if (showsAdamsPose) .025f else .07f), Offset(size.width * .5f, hipY), body.centered)
                line(ls, rs, body.shouldersReady)
                line(ls, Offset(size.width * if (isAdams) .49f else .24f, size.height * if (isAdams && showsAdamsPose) .70f else .50f), body.shouldersReady)
                line(rs, Offset(size.width * if (isAdams) .51f else .76f, size.height * if (isAdams && showsAdamsPose) .70f else .50f), body.shouldersReady)
                line(lh, rh, body.hipsReady)
                if (task == BodyCaptureTask.Seated) {
                    line(lh, Offset(size.width * .35f, size.height * .88f), false)
                    line(rh, Offset(size.width * .65f, size.height * .88f), false)
                } else {
                    line(lh, lk, body.kneesReady); line(rh, rk, body.kneesReady)
                    line(lk, lf, body.feetReady); line(rk, rf, body.feetReady)
                }
                listOf(head to body.headReady, ls to body.shouldersReady, rs to body.shouldersReady, lh to body.hipsReady, rh to body.hipsReady, lk to body.kneesReady, rk to body.kneesReady, lf to body.feetReady, rf to body.feetReady).forEach { (point, ready) ->
                    drawCircle(if (ready) Color(0xFF28C76F) else Color(0xFFFFD54F), 5.dp.toPx(), point)
                }
                }
            }
            Surface(color = Color.Black.copy(alpha = .52f), shape = RoundedCornerShape(50)) {
                Text(
                    if (task == BodyCaptureTask.FootArch) {
                        if (recording) "正在记录足部近景：保持双脚平行站稳" else "仅将双膝以下、双踝、足跟和双脚放入近景框"
                    } else if (recording) {
                        if (task == BodyCaptureTask.ForwardBend) "对准前屈轮廓：双脚并拢、双膝伸直、头部放松" else "正在记录：保持手机固定，按语音完成动作"
                    } else "将头、肩、髋、膝和双脚对准标定点",
                    color = if (body.isReady) Color(0xFF55E392) else Color.White,
                    fontSize = 13.sp,
                    modifier = Modifier.padding(horizontal = 12.dp, vertical = 7.dp)
                )
            }
        }
    }
}

@Composable
private fun DeviceLevelBubble(alignment: CaptureDeviceAlignment) {
    Surface(color = Color.Black.copy(alpha = .52f), shape = RoundedCornerShape(50)) {
        Row(Modifier.padding(horizontal = 12.dp, vertical = 7.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Box(Modifier.size(34.dp).background(Color.White.copy(alpha = .10f), CircleShape), contentAlignment = Alignment.Center) {
                Canvas(Modifier.fillMaxSize()) {
                    drawLine(Color.White.copy(alpha = .45f), Offset(size.width / 2, 3f), Offset(size.width / 2, size.height - 3f), 1f)
                    drawLine(Color.White.copy(alpha = .45f), Offset(3f, size.height / 2), Offset(size.width - 3f, size.height / 2), 1f)
                    drawCircle(
                        if (alignment.isLevel) Color(0xFF28C76F) else Color(0xFFFFD54F),
                        4.5.dp.toPx(),
                        Offset(size.width / 2 + alignment.rollDegrees.coerceIn(-10.0, 10.0).toFloat() * 1.2f, size.height / 2 + alignment.pitchDegrees.coerceIn(-10.0, 10.0).toFloat() * 1.2f)
                    )
                }
            }
            Column {
                Text(if (alignment.isLevel) "手机角度已校正" else "请调平手机", color = Color.White, fontSize = 13.sp)
                Text(if (alignment.available) "左右 ${alignment.rollDegrees.toInt()}° · 前后 ${alignment.pitchDegrees.toInt()}°" else "正在读取陀螺仪", color = Color.White.copy(alpha = .72f), fontSize = 12.sp)
            }
        }
    }
}

@Composable
fun CaptureCalibrationStatusRow(cameraReady: Boolean, deviceReady: Boolean, bodyReady: Boolean) {
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        CalibrationStatusItem("相机", cameraReady, Modifier.weight(1f), true)
        CalibrationStatusItem("手机角度", deviceReady, Modifier.weight(1f), false)
        CalibrationStatusItem("人体对齐", bodyReady, Modifier.weight(1f), false)
    }
}

@Composable
private fun CalibrationStatusItem(title: String, ready: Boolean, modifier: Modifier, camera: Boolean) {
    val icon = if (ready) Icons.Filled.CheckCircle else if (camera) Icons.Filled.CameraAlt else if (title == "手机角度") Icons.Filled.ScreenRotation else Icons.Filled.AccessibilityNew
    Row(modifier.heightIn(min = 32.dp).background(Color.White.copy(alpha = .07f), RoundedCornerShape(50)).padding(horizontal = 7.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.Center) {
        Icon(icon, null, tint = if (ready) Color(0xFF55E392) else Color.White.copy(alpha = .72f), modifier = Modifier.size(15.dp))
        Spacer(Modifier.width(3.dp)); Text(title, color = if (ready) Color(0xFF55E392) else Color.White.copy(alpha = .72f), fontSize = 12.sp)
    }
}
